src/openzeppelin/token/erc721/library.cairo

```diff
@@ -14,549 +14,461 @@ from openzeppelin.introspection.erc165.IERC165 import IERC165
+        // then this method should fail.
+        // The `caller` address and `is_approved` boolean are both field elements
+        // meaning that a*0==0 for all a in the field,
+        // therefore a*b==0 implies that at least one of a,b is zero in the field
+
+        _transfer(from_, to, token_id);
+        return ();
+    }
+
+    func safe_transfer_from{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_check_ptr}(
+        from_: felt, to: felt, token_id: Uint256, data_len: felt, data: felt*
+    ) {
+        alloc_locals;
+        with_attr error_message("ERC721: token_id is not a valid Uint256") {
+            uint256_check(token_id);
+        }
+        let (caller) = get_caller_address();
+        let is_approved = _is_approved_or_owner(caller, token_id);
+        with_attr error_message(
+                "ERC721: either is not approved or the caller is the zero address") {
+            assert_not_zero(caller * is_approved);
+        }
+        // Note that if either `is_approved` or `caller` equals `0`,
+        // then this method should fail.
+        // The `caller` address and `is_approved` boolean are both field elements
+        // meaning that a*0==0 for all a in the field,
+        // therefore a*b==0 implies that at least one of a,b is zero in the field
+
+        _safe_transfer(from_, to, token_id, data_len, data);
+        return ();
+    }
+
+    //
+    // Internals
+    //
+
+    func assert_only_token_owner{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_check_ptr}(
+        token_id: Uint256
+    ) {
+        uint256_check(token_id);
+        let (caller) = get_caller_address();
+        let (owner) = owner_of(token_id);
+        // Note `owner_of` checks that the owner is not the zero address
+        with_attr error_message("ERC721: caller is not the token owner") {
+            assert caller = owner;
+        }
+        return ();
+    }
+
+    func _is_approved_or_owner{pedersen_ptr: HashBuiltin*, syscall_ptr: felt*, range_check_ptr}(
+        spender: felt, token_id: Uint256
+    ) -> felt {
+        alloc_locals;
+
+        let exists = _exists(token_id);
+        with_attr error_message("ERC721: token id does not exist") {
+            assert exists = TRUE;
+        }
+
+        let (owner) = owner_of(token_id);
+        if (owner == spender) {
+            return TRUE;
+        }
+
+        let (approved_addr) = get_approved(token_id);
+        if (approved_addr == spender) {
+            return TRUE;
+        }
+
+        let (is_operator) = is_approved_for_all(owner, spender);
+        if (is_operator == TRUE) {
+            return TRUE;
+        }
+
+        return FALSE;
+    }
+
+    func _exists{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        token_id: Uint256
+    ) -> felt {
+        let (exists) = ERC721_owners.read(token_id);
+        if (exists == FALSE) {
+            return FALSE;
+        }
+
+        return TRUE;
```

Comment:
```suggestion
        return exists;
```
?

or directly returning the `read` if we change the return tyoe of the storage var

Reply:
With this func, the `exists` variable returns the tokenId's owner address (or `0` if none). If we follow the suggestion, we'd have to refactor the assertions of the functions that call `_exists()`. This is how it's used now:

```
        let exists = _exists(token_id);
        with_attr error_message("ERC721: approved query for nonexistent token") {
            assert exists = TRUE;
        }
```

It's not a big deal to change, but I think the suggestion conflicts with the spirit of `_exists`, no?

Reply:
right. ignore me.